platform/include/platform/CircularBuffer.h

```diff
@@ -98,63 +98,204 @@ class CircularBuffer {
+     *
+     * @param dest The array which will receive the elements.
+     * @param len The number of elements to pop.
+     *
+     * @return The number of elements popped.
+     */
+    CounterType pop(T* dest, CounterType len)
+    {
+        CounterType data_popped = 0;
+
+        core_util_critical_section_enter();
+
+        if(!non_critical_empty()) {
+            /* make sure we only try to read as much as we have items present */
+            if (len > non_critical_size()) {
+                len = non_critical_size();
             }
+            data_popped = len;
+
+            /* items may be split by overlap, take only the number we have to the right of tail */
+            if ((_tail + data_popped) > BufferSize) {
+                data_popped = BufferSize - _tail;
+            }
+
+            std::copy(_buffer + _tail, _buffer + _tail + data_popped, dest);
+
+            /* if we looped over the end we may need to pop again */
+            CounterType left_to_pop = len - data_popped;
+
+            if (left_to_pop) {
+                std::copy(_buffer, _buffer + left_to_pop, dest + data_popped);
+
+                data_popped += left_to_pop;
+            }
+
+            _tail += data_popped;
+            _tail %= BufferSize;
             _full = false;
-            data_popped = true;
         }
+
         core_util_critical_section_exit();
+
         return data_popped;
     }
 
-    /** Check if the buffer is empty
+    /**
+     * Pop multiple elements from the buffer.
+     *
+     * @param dest The array which will receive the elements.
+     *
+     * @return The number of elements popped.
+     */
+    template<CounterType N>
+    CounterType pop(T (&dest)[N])
+    {
+        return pop(dest, N);
+    }
+
+    /**
+     * Pop multiple elements from the buffer.
+     *
+     * @param dest The span that contains the buffer that will be used to store the elements.
      *
-     * @return True if the buffer is empty, false if not
+     * @return The span with the size set to number of elements popped using the buffer passed in as the parameter.
+     */
+    mbed::Span<T> pop(mbed::Span<T>& dest)
+    {
+        CounterType popped = pop(dest.data(), dest.size());
+        return mbed::make_Span(dest.data(), popped);
+    }
+
+    /** Check if the buffer is empty.
+     *
+     * @return True if the buffer is empty, false if not.
      */
     bool empty() const
     {
         core_util_critical_section_enter();
-        bool is_empty = (_head == _tail) && !_full;
+        bool is_empty = non_critical_empty();
         core_util_critical_section_exit();
         return is_empty;
     }
 
-    /** Check if the buffer is full
+    /** Check if the buffer is full.
      *
      * @return True if the buffer is full, false if not
      */
     bool full() const
     {
-        core_util_critical_section_enter();
-        bool full = _full;
-        core_util_critical_section_exit();
-        return full;
+        return _full;
```

Comment:
I think you want this to be at least an atomic load. Otherwise existing icky spin code like

    `while (buff.full()) { }`

waiting for an interrupt to drain the buffer will fail.

Indeed, that's the implementation I see over here on this project branch.

Reply:
thanks, made atomic